contracts/src/cryptography/eip712.rs

```diff
@@ -0,0 +1,246 @@
+//! [EIP-712](https://eips.ethereum.org/EIPS/eip-712) is a standard for hashing
+//! and signing typed structured data.
+//!
+//! The implementation of the domain separator was designed to be as efficient
+//! as possible while still properly updating the chain id to protect against
+//! replay attacks on an eventual fork of the chain.
+//!
+//! NOTE: This contract implements the version of the encoding known as "v4", as
+//! implemented by the JSON RPC method [`eth_signTypedDataV4`] in MetaMask.
+//!
+//! [`eth_signTypedDataV4`]: https://docs.metamask.io/guide/signing-data.html
+
+use alloc::{string::String, vec::Vec};
+
+use alloy_primitives::{
+    b256, fixed_bytes, keccak256, Address, FixedBytes, B256, U256,
+};
+use alloy_sol_types::{sol, SolType};
+use stylus_sdk::{
+    block, contract,
+    stylus_proc::{external, sol_storage},
+};
+
+use super::message_hash_utils::to_typed_data_hash;
+
+/// `keccak256("EIP712Domain(string name,string version,uint256 chainId,address
+/// verifyingContract)");`
+pub const TYPE_HASH: B256 =
+    b256!("8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f");
+pub const FIELDS: FixedBytes<1> = fixed_bytes!("15");
+pub const SALT: B256 = B256::ZERO;
+
+pub type DomainSeparatorTuple = sol! {
+    tuple(bytes32, bytes32, bytes32, uint256, address)
+};
+
+sol_storage! {
+    /// State of an `Eip712` contract.
+    pub struct Eip712 {
+        /// The cached domain separator.
+        bytes32 _cached_domain_separator;
+        /// The cached chain id.
+        uint256 _cached_chain_id;
+        /// The cached contract address, i.e. `address(this)`.
+        address _cached_this;
+        /// The hashed name.
+        bytes32 _hashed_name;
+        /// The hashed version.
+        bytes32 _hashed_version;
+        /// The name.
+        string _name;
+        /// The version.
+        string _version;
+    }
+}
```

Comment:
In the Solidity implementation, you can see that most state variables are in reality private immutables, set during contract construction:

```solidity
bytes32 private immutable _cachedDomainSeparator;
uint256 private immutable _cachedChainId;
address private immutable _cachedThis;

bytes32 private immutable _hashedName;
bytes32 private immutable _hashedVersion;

ShortString private immutable _name;
ShortString private immutable _version;
string private _nameFallback;
string private _versionFallback;

constructor(string memory name, string memory version) {
    _name = name.toShortStringWithFallback(_nameFallback);
    _version = version.toShortStringWithFallback(_versionFallback);
    _hashedName = keccak256(bytes(name));
    _hashedVersion = keccak256(bytes(version));

    _cachedChainId = block.chainid;
    _cachedDomainSeparator = _buildDomainSeparator();
    _cachedThis = address(this);
}
```

In reality, the only storage used is `_nameFallback` and `_versionFallback`. I think we can probably remove *all* that state:

- private immutables are just search and replace in the compiled binary. They are equivalent to Rust's `const` variables.
- `_nameFallback` and `_versionFallback` are only used in case the consumer doesn't properly set `name` and `version`, but if they are set at compile time, then we don't need them at all.

It turns out this EIP implementation doesn't even have to be a contract, since it doesn't need storage -- we can resolve these at compile time by making the user specify the variables. Something like the [ERC-20 example](https://github.com/OffchainLabs/stylus-sdk-rs/blob/17fc86a4615af0769e8f528f32276e067702d52e/examples/erc20/src/lib.rs#L21-L24) of the SDK, but not even requiring it to be a contract, just a plain ol' Rust generic trait.

Let's do this and move all of this code to `lib/crypto` please.

Reply:
please correct me if I am wrong; 


This means we would rather introduce a `EIP712` trait, having the `eip712_domain` method;

```rust 
pub fn eip712_domain(
        &self,
    ) -> (FixedBytes<1>, String, String, u64, Address, B256, Vec<U256>);
pub fn eip712_name(&self) -> String;
pub fn eip712_version(&self) -> String;
```

and the rest of the methods would be extracted to be standalone; 
- `build_domain_separator`
- `domain_separator_v4`
- `hash_typed_data_v4`

lastly, the `NAME` and `VERSION` would be associated const types.

cc: @alexfertel

Reply:
please I need your assertion on this, @alexfertel

Reply:
I think we only need one `eip712_domain`, cause that's the one which may require a custom implementation. Everything else can be associated consts. I may be wrong though! I'd need to see the actual implementation to know.

Correct me if I'm wrong, but we can compute everything at compile time based on the consumer's implementation. Let's give that I try. If you need a keccak at compile time, you can use `const_keccak`.

Reply:
From what I can see it looks like it should be doable and it would be a good gas optimization when having everything calculated at compile time. With this attitude, we would need to have integration tests (E2E) confirming that we are able to set & calculate everything in "dummy" parent contract.

Reply:
What I can see in [ERC-20 Permit extension](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC20Permit.sol), we need:
- `domain_separator_v4`
- `hash_typed_data_v4`

Reply:
If we have everything ready at compile-time, we don't need those, since the first one only checks for the cached values (which will be there since we compute them at compile-time) and the second one is this:

```rust
        assembly {
            let ptr := mload(0x40)
            mstore(ptr, hex"19_01")
            mstore(add(ptr, 0x02), domainSeparator)
            mstore(add(ptr, 0x22), structHash)
            digest := keccak256(ptr, 0x42)
        }
```

which you can see is `keccak256([&[0x19, 0x01], &domain_hash[..], &struct_hash[..]].concat())`, which we could do at compile-time using `const_keccak`.

I may be wrong though! If at any point you find issues, let me know.